### PR TITLE
Using "virtual padding" to calculate number_local_tokens per SP rank, and fix  `combine()` shape mismatch

### PR DESCRIPTION
**Pull request by wwwjn**
## What's the problem
- [Currently] combine() wrongly assume the tokens are evenly sharded on each rank https://github.com/pytorch/torchtitan/blob/c0428bb186f5c97e1d1b4ed89febd22916eee302/torchtitan/models/common/token_dispatcher.py#L439-L456  (infer global SPMD in local SPMD region)
    - If uneven sharded, out_TD will have different shapes across SP ranks.
    - We should directly ban if input number of tokens in input batch can not be evenly sharded by SP ranks

- [Future] Router will use spmd_types soon, and router is per SP rank. Per SP rank should have even sharding
- [Future] we want to avoid dispatch/load_balacing the padded token, we should be able to do that by adding metadata field to record the actually local tokens for each sp rank 

## What does this PR do?

This PR is doing "virtual padding" , and passing metadata around
- Calculate num_local_tokens_after_padding = (T + pad_tokens) // sp_size in MoE module
- Pass num_local_tokens_after_padding to GroupedExperts module, then to combine()
- combine() returns a tensor with shape (num_local_tokens_after_padding * sp_rank, .... )
- slice the combined tensor to (T, ...)  in MoE

### CODE REVIEW

File: torchtitan/models/common/token_dispatcher.py [hunk 1]
```diff
@@ -55,6 +67,10 @@ def __init__(self, config: Config):
         self.num_experts = config.num_experts
         self.top_k = config.top_k
         self.score_before_experts = config.score_before_experts
+        # Sequence-parallel split coordinates. EP dispatchers update these in
+        # wire_meshes(); the local dispatcher keeps the TP=1 defaults.
```
Comment (wwwjn) [inline torchtitan/models/common/token_dispatcher.py]: Introducing this sp_rank, sp_size in LocalTokenDispatcher is not ideal, but it's used in local_num_valid_tokens() , I want to share local_num_valid_tokens() implementation across AllToAllTokenDispather, DeepEP/HybridEP

File: torchtitan/models/common/token_dispatcher.py [hunk 2]
```diff
@@ -65,6 +81,28 @@ def wire_meshes(
         """No-op for the EP=1 dispatcher. Subclasses override."""
         del ep_mesh, tp_mesh
 
+    def local_num_valid_tokens(
```
Comment (wwwjn) [inline torchtitan/models/common/token_dispatcher.py]: Ideally we should move to All2AllTokenDipsatcher / DeepEP / HybridEPTokenDispatcher, will decide a better place

File: torchtitan/models/common/moe.py [hunk 3]
```diff
@@ -368,6 +401,20 @@ def forward(self, x_BLD: torch.Tensor) -> torch.Tensor:
...
+        moe_padding = None
+        sp_size = getattr(self.experts.token_dispatcher, "sp_size", 1)
+        pad_tokens = (-T) % sp_size
+        if pad_tokens:
+            # Record the padded global length before routing, but do not
+            # physically pad this DTensor. Cat along Shard(1) would unshard the
+            # sequence dimension. Padding is logically appended to the flattened
+            # global token tail; GroupedExperts materializes only each local
+            # shard's tail padding.
+            moe_padding = MoEPaddingMetadata(
```
Comment (wwwjn) [inline torchtitan/models/common/moe.py]: All these metadata fields are actually integer, not a tensor. So these won't create D2H sync

File: torchtitan/models/common/decoder.py [hunk 4]
```diff
@@ -200,6 +201,31 @@ def __init__(self, config: Config):
         self.norm = config.norm.build()
         self.lm_head = config.lm_head.build()
 
+    @staticmethod
+    def _validate_sequence_parallel_input(x: torch.Tensor) -> None:
```
Comment (tianyu-l) [inline torchtitan/models/common/decoder.py]: Can we do this in `Decoder.update_from_config()`, after sharding is set?
E.g. in each model, we can call `Decoder.update_form_config` after model-specific udpate/check is done, instead of before https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/llama3/model.py#L72. Not sure if it's compatible with other checks.

File: torchtitan/models/common/token_dispatcher.py [hunk 5]
```diff
@@ -433,11 +450,11 @@ def combine(
...
 
-        # With SP, x_TD is the local shard. Create full-size buffer for
-        # scatter_add so routed results from all SP ranks can be placed
-        # at global positions.
+        # With SP, create a full-size buffer for scatter_add so routed results
+        # from all SP ranks can be placed at global positions. Padded tail rows
+        # are never routed and are sliced off below.
         out_TD = torch.zeros(
-            x_TD.shape[0] * self.sp_size,
+            num_local_tokens_after_padding * self.sp_size,
```
Comment (tianyu-l) [inline torchtitan/models/common/token_dispatcher.py:461]: My impression is that this "virtual padding" idea would work when `GroupedExperts` is in a local region inside a DTensor global region, where DTensor handles the shard / all-gather.

After we migrate to spmd_types, the all-gather would require this metadata as well (`spmd.redistribute` takes this shape arg), with which spmd_types could also achieve similar "pad / unpad only around collectives" effect. O/w we still have to do "real padding" in model code. cc @pianpwk 

I'm OK with this temporary solution (after cleanup) to unblock your vLLM + MoE work.

File: torchtitan/models/common/moe.py [hunk 6]
```diff
@@ -368,6 +379,14 @@ def forward(self, x_BLD: torch.Tensor) -> torch.Tensor:
         the GroupedExperts boundary.
         """
         B, L, D = x_BLD.shape
+        T = B * L
+        sp_size = getattr(self.experts.token_dispatcher, "sp_size", 1)
+        pad_tokens = (-T) % sp_size
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py:384]: > We never materialize a padded token , just calculating the shapes that needs to be returned in combine(). The hidden assumption is "padding always happens on the last SP rank"

Why do you say "padding always happens on the last SP rank"?

In an extremal case, what if sp_size = 1024, but T = 1?

Comment (wwwjn) [reply torchtitan/models/common/moe.py:384]: > padding always happens on the last SP rank"

You are right, the description is not accurate. It should be always happening at the end of sequence (tail padding), so we can infer the start / end index on each local rank.

File: torchtitan/models/common/token_dispatcher.py (hunk 5 above)
Comment (wwwjn) [reply torchtitan/models/common/token_dispatcher.py:461]: Yes let me clean up. 

For spmd_types in MoE region, we don't need "real padding" in our current setup, just passing this "vitural padded shape" as metadata around. https://github.com/pytorch/torchtitan/blob/da230bbb205cd60ea30d485cb359bed2830adb84/torchtitan/models/common/MOE_SHARDING.md#L4

File: torchtitan/models/common/decoder.py (hunk 4 above)
Comment (wwwjn) [reply torchtitan/models/common/decoder.py]: I was thinking there are 2 ways to guard this (This part is not good and needs iterate)
1. check config `TrainingConfig.seq_len` 
2. check real input `x`'s  shape

If we use 1, RL's PolicyTrainer path would also need to use the TrainingConfig https://github.com/pytorch/torchtitan/blob/7f5d1193299edae528064e13ac4faa5a94f31e68/torchtitan/experiments/rl/actors/trainer.py#L303  but today, the input size to RL's PolicyTrainer is decided by Batcher's `BatchConfig.seq_length`, (`TrainingConfig` it not clearly used in RL's PolicyTrainer, which needs to be cleared), that why I thought 2 is straight-forward.

I think we can do following:
1. In pretraining: We check TrainerConfig.seq_len in `update_from_config()`, and move update_from_config after set_xx_sharding_config
3. In RL, check BatchConfig.seq_len can be divided by sp_degree in `RLTrainer.__post_init__ ` (not PolicyTrainer)

Comment (wwwjn) [reply torchtitan/models/common/decoder.py]: > in each model, we can call `Decoder.update_form_config` after model-specific udpate/check is done, instead of before 

Yes I can add that to current PR

**Comment by wwwjn** [comment]
Close this because of #3595 . Let's move discussion there

**Review by tianyu-l — COMMENTED** [review]
sounds right to me, one concrete issue before landing
